### PR TITLE
SST reservoir training validation for sweeps

### PR DESCRIPTION
**Pull request by frodre**
This PR adds a secondary style of validation specific to the SST reservoirs to the reservoir training.  It borrows from Paula's validation style but implements it to be compatible with wandb sweeps.  Currently, it is set to track the one-step and rollout RMSE (added together) as the primary metric to minimizae.  Other outputs include spatial measures of bias, MAE, RMSE, MSE, and skill compared against one-step persistence for both one-step and rollouts.

TODO: 
- [ ] add a link to an example wandb sweep
- [ ] Describe new single-file NetCDF to tf dataset loader
- [ ] New method for setting state (instead of just being able to load a saved state on init)
- [ ] Adds overlap trimming for hybrid mask (in case of reservoir inputs w/ overlap)
- [ ] Describe Sweep configuration and usage

Coverage reports (updated automatically):

### CODE REVIEW

File: external/fv3fit/fv3fit/reservoir/train.py [hunk 1]
```diff
@@ -78,10 +80,23 @@ def _get_transformers(
     return TransformerGroup(**transformers)
 
 
+def _expand_mask_zdim(mask: tf.Tensor, z_dim_len: int) -> tf.Tensor:
```
Comment (frodre) [inline external/fv3fit/fv3fit/reservoir/train.py:83]: Is this used?

File: external/fv3fit/fv3fit/reservoir/reservoir.py [hunk 2]
```diff
@@ -75,7 +75,9 @@ def increment_state(self, input):
             masked_input = input * self.input_mask_array
         else:
             masked_input = input
-        self.state = np.tanh(masked_input @ self.W_in.T + self.state @ self.W_res.T)
+        self.state: np.ndarray = np.tanh(
```
Comment (frodre) [inline external/fv3fit/fv3fit/reservoir/reservoir.py:78]: Typing necessary with new `set_state` method

File: external/fv3fit/fv3fit/reservoir/config.py [hunk 3]
```diff
@@ -106,6 +106,7 @@ class ReservoirTrainingConfig(Hyperparameters):
     square_half_hidden_state: bool = False
     hybrid_variables: Optional[Sequence[str]] = None
     mask_variable: Optional[str] = None
+    validate_sst_only: bool = False
```
Comment (frodre) [inline external/fv3fit/fv3fit/reservoir/config.py:109]: New flag that switches between atmosphere validation and validation I've been using for sst models.  Defaults to `False` for backwards compatibility

File: external/fv3fit/fv3fit/reservoir/utils.py [hunk 4]
```diff
@@ -26,10 +27,14 @@ def assure_txyz_dims(var_data: tf.Tensor) -> tf.Tensor:
     elif len(var_data.shape) == 3:
         orig_shape = var_data.shape
         reshaped_tensor = tf.reshape(var_data, shape=(*orig_shape, 1))
+    elif len(var_data.shape) == 2:
```
Comment (frodre) [inline external/fv3fit/fv3fit/reservoir/utils.py:30]: When was 2D data relevant?

File: external/fv3fit/tests/training/test_reservoir.py [hunk 5]
```diff
@@ -34,7 +34,8 @@ def test_train_reservoir():
         .transpose("time", "x", "y", "z")
     )
     train_tfdataset = tfdataset_from_batches([train_dataset for _ in range(4)])
-    val_tfdataset = tfdataset_from_batches([test_dataset])
```
Comment (frodre) [inline external/fv3fit/tests/training/test_reservoir.py:37]: Fix?

**Comment by frodre** [comment]
Closed for #2383
